src/com/amazon/ion/impl/ResizingPipedInputStream.java

```diff
@@ -0,0 +1,557 @@
+     * the caller should ensure that the given `InputStream` has at least `len` bytes available before calling
+     * this method or provide an InputStream implementation that does not block.
+     * @param input the source of the bytes.
+     * @param len the number of bytes to attempt to write.
+     * @return the number of bytes actually written, which will only be less than `len` if
+     * {@link InputStream#read(byte[], int, int)} returns less than `len`.
+     * @throws IOException if thrown by the given `InputStream` during read, except for {@link EOFException}. If an
+     *   EOFException is thrown by the `InputStream`, it will be caught and this method will return the number of bytes
+     *   that were received before the exception was thrown.
+     */
+    public int receive(final InputStream input, final int len) throws IOException  {
+        ensureSpaceInBuffer(len);
+        int numberOfBytesRead;
+        try {
+            numberOfBytesRead = input.read(buffer, writeIndex, len);
+        } catch (EOFException e) {
+            // Some InputStream implementations (such as GZIPInputStream) will throw EOFException instead of
+            // returning -1.
+            numberOfBytesRead = -1;
+        }
+        if (numberOfBytesRead > 0) {
+            writeIndex += numberOfBytesRead;
+            size += numberOfBytesRead;
+        } else {
+            numberOfBytesRead = 0;
+        }
+        if (!useBoundary) {
+            extendBoundary(numberOfBytesRead);
+        }
+        return numberOfBytesRead;
+    }
+
+    /**
+     * {@inheritDoc}
+     * <p>
+     * NOTE: This method adheres to the documented behavior of {@link InputStream#read(byte[], int, int)}
+     * <strong>except</strong> that it never blocks. If a read is attempted before the first write,
+     * this method will return -1.
+     */
+    @Override
+    public int read(final byte[] b, final int off, final int len) {
+        if (b.length == 0 || len == 0) {
+            return 0;
+        }
+        if (available < 1) {
+            return -1;
+        }
+        int bytesToRead = Math.min(available, len);
+        System.arraycopy(buffer, readIndex, b, off, bytesToRead);
+        readIndex += bytesToRead;
+        available -= bytesToRead;
+        size -= bytesToRead;
+        return bytesToRead;
+    }
+
+    /**
+     * Copies all of the available bytes in the buffer without changing the number of bytes available to subsequent
+     * reads.
+     * @param outputStream stream to which the bytes will be copied.
+     * @throws IOException if thrown by {@link OutputStream#write(byte[], int, int)}.
+     */
+    public void copyTo(final OutputStream outputStream) throws IOException {
+        outputStream.write(buffer, readIndex, available);
+    }
+
+    /**
+     * Seeks the read index to the given position.
+     * @param index the index to which to seek.
+     */
+    void seekTo(int index) {
```

Comment:
No, but since this isn't a public method I decided not to add a bounds check to every invocation. I'm happy to update the javadoc to call out that the index must be positive, though.